wcコマンドを作るの提出物です

rira100000000 opened the pull request:
[wc コマンドを作る](https://bootcamp.fjord.jp/practices/161)の提出物です。

rira100000000 committed:
first commit

rira100000000 committed:
wcコマンドの実装

wcコマンドでオプションを使用しない場合の処理を実装した

rira100000000 committed:
totalを表示する機能を追加

複数のパスを与えられたとき、totalを表示する機能を追加
付随してテストを作成

rira100000000 committed:
テストの追加

ファイルが存在しない場合のテストを追加
ディレクトリを指定した場合のテストを追加

rira100000000 committed:
rubocop対応

rira100000000 committed:
rubocop対応

print_all_countの処理の一部をfill_count_listとfill_total_hashに分割
make_display_lineの処理の一部をfill_counts_by_zeroに分割

rira100000000 committed:
オプション機能の追加

オプションl,w,cを指定した場合の処理を追加
一部メソッドの削除、名前の変更
一部処理をmake_display_lineとmake_display_total_lineにメソッド化

rira100000000 committed:
パイプを受け取る機能を追加

パイプを受け取った場合の処理を追加
付随してテストの追加

rira100000000 committed:
標準入力への対応

標準入力に対応するよう修正

rira100000000 committed:
不具合修正

パスを指定して実行するとファイル名が出ないバグを修正
requireされた先で実行しないように修正

rira100000000 committed:
rubocop対応

rira100000000 committed:
誤字修正

maedana commented on 05.wc/test/wc_test.rb:
```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
```
テストコードかいてあっていいですね :+1:

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'debug'
```
デバッグ用だと思いますので提出コードからは消しましょうー

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+    options[:l] = true
+    options[:w] = true
+    options[:c] = true
+  end
```
初期値は`parse_option`に閉じ込めるほうが役割がわかりやすいかなと思いました。いかがでしょうか?

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+    options[:w] = true
+    options[:c] = true
+  end
+  if paths != []
```
if/else両方ありますし否定条件は使わず、先に肯定条件となる標準入力用のロジックを書いたほうが読みやすいかもですね

それと配列の空判定は https://docs.ruby-lang.org/ja/latest/method/Array/i/empty=3f.html というのを使うとよりリーダブルかもしれませんね

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+  else
+    while (line = gets)
+      standard_input << line
+    end
```
https://docs.ruby-lang.org/ja/latest/class/Kernel.html#M_READLINES

を使うとループしなくても一気に読み込めてシンプルなコードになりそうですね

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+end
+
+def print_count(paths, options)
+  max_length_hash = reset_max_length_hash
```
total_hashの初期化はメソッドにしてませんし、こっちも単に変数の初期化ですしメソッドにしなくてもいいんじゃないですかね?

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+  total_hash
+end
+
+def print_count(paths, options)
```
この命名だと凄く汎用的にみえますけど実際はパス指定用ですよね。なので`print_standard_input_count`と対になるような命名のほうがわかりやすいのかなって思いました。この命名だと標準入力だろうがファイル入力だろうがいい感じに出してほしいように感じました。

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+      count_list[index][:directory] = "wc: #{path}: Is a directory"
+    elsif File.exist?(path)
+      count_list = fill_count_list(count_list, index, path)
+      total_hash = fill_total_hash(total_hash, count_list, index)
```
数字を埋める処理をやっていてそれ自体はいいのですが、一方で`File.directory?(path)`のパターンでは、`make_display_str`内で`fill_counts_by_zero`で初期値入れてるんですよね。`File.directory?(path)`のパターンでも初期値はこのメソッド内で入れてあげるほうが一貫性があって読みやすいように感じます。いかがでしょうか?

maedana replied:
`fill_counts_by_zero`, `fill_count_list`, `fill_total_hash`がいずれも引数を改変する`副作用のあるメソッド`になってます。オブジェクト指向でインスタンス変数の値を書き換えるならいいんですが、引数の値を書き換えるのはできれば避けたほうがいいと思うんですよね。

今回のケースですとようは`count_list[index]`に適切なHashオブジェクトを代入したいというのが目的かと思いました。代入はあくまでも呼び出し元であるこのメソッド内でやるようにして、例えば`fill_count_list`のようなものは`path`だけを引数にして`number_of_lines`, `number_of_words`, `number_of_bytes`の3つの値を戻り値にするようなメソッドにしてその値を利用して`count_list[index]`のHashを構築するような形にするのが自然かなと。いかがでしょうか?(メソッド名も`fill`ではなくなるので違う命名にしないといけないと思います)

maedana commented on Gemfile:
```diff
@@ -0,0 +1,8 @@
+
+source "https://rubygems.org"
+
+gem "rails"
```
railsがあるのは変ですね

maedana commented on Gemfile:
```diff
@@ -0,0 +1,8 @@
+source "https://rubygems.org"
+
+gem "rails"
+gem "rubocop"
```
rubocop-fjordの依存で入るので不要ですね

maedana commented on Gemfile:
```diff
@@ -0,0 +1,8 @@
+gem "rails"
+gem "rubocop"
+gem "debug"
+gem 'rubocop-fjord', '~> 0.2.2'
```
debugとrubocop-fjordは`gem install xxx`で入れたら使えるでしょうし、そうなると`Gemfile`自体なくていいんじゃないかと思いました。いかがでしょう?

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+def make_display_str(counts_list, max_length_hash, total_hash, options)
+  result = counts_list.map do |counts|
+    if counts[:file_not_exists]
+      counts[:file_not_exists]
```
`file_not_exists`をセットしている箇所がないように見えます。この分岐不要なのでは?

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,141 @@
+        display_str = +"#{counts[:directory]}\n"
+      elsif counts[:not_file]
+        next display_str = +"#{counts[:not_file]}\n"
+      end
```
この辺の分岐を`print_count`でもやってるんですよね。`print_count`のループ内で随時`make_display_line`を使って出力していけばこのメソッドで再びループしなくて済みそうに見えました。すると似たような分岐が複数メソッドに散らばることもなく、このメソッドもなくせてスッキリできそうに見えますがいかがでしょう?

rira100000000 replied:
`print_count`で、最終的に表示する項目の最大文字数を`max_length_hash`に格納しています。
この最大文字数を調べるためにはtotalで表示される値を知る必要があるので、totalを計算する回と表示用に`make_display_line`を実行する回でループが2回必要になると思うのですが、どうでしょうか？

maedana replied:
確かにおっしゃる通りですね。であれば今のままでOKです :+1:

rira100000000 committed:
Gemfileの削除

Gemfileを使う必要がなかったため削除しました

rira100000000 committed:
require debugを削除

rira100000000 committed:
オプションが空の場合の処理を移動

optionsが空の場合の処理をmainからparse_optionに移動しました

rira100000000 committed:
mainのリファクタリング

mainの処理をリファクタリングしました

rira100000000 committed:
mainのリファクタリング

ループ処理をreadlinesに変更しました

rira100000000 committed:
不要なメソッドの削除

メソッドにする必要がなかったので、reset_max_length_hashを削除しました

rira100000000 committed:
引数を改変する処理の修正

引数を改変する処理を削除し、必要な値を返す処理に変更しました

rira100000000 committed:
ディレクトリが指定された場合の処理を移動

処理に一貫性を持たせるため、make_display_strで行っていたディレクトリが指定された場合の処理をprint_command_line_argument_countに移動しました

rira100000000 committed:
不要な分岐の削除

必要のない分岐を削除しました

rira100000000 committed:
ループ不要の処理をループ外へ移動

max_length_hashへの値の格納はループする必要がなかったため、ループの外に移動しました

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,122 @@
+  stat.size
+end
+
+def return_zeros_for_counts
```
ここに限らずですけど、`return_`はなくていいかなーと思います。メソッドの原則は動詞始まりなので間違いじゃないんですが、Rubyの文化的に単に値を返すメソッドのときは動詞を省略するのはよくあるかなーと。`File#size`のようなメソッドも`File#return_size`とはなってないですよね。

それとここに関してはただの固定値ですしメソッドにまでしなくていいかなって思いました。

rira100000000 replied:
メソッドを削除しました。
1行で書くとrubocopで「Do not use parallel assignment」に引っかかるので3行で書いています。
```
count_list[index][:number_of_lines] = 0
count_list[index][:number_of_words] = 0
count_list[index][:number_of_bytes] = 0
```

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,122 @@
+  result
+end
+
+def return_count_list(path)
```
同上です。return_なくていいのかなって思います。

rira100000000 replied:
count_listだとすでに配列名で使っていたのでlist_countsに変更しました。

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,122 @@
+  [count_lines(str), count_words(str), count_bytes(path)]
+end
+
+def return_total_list(total_hash, count_list, index)
```
同上です。あと後述しますけどこのメソッドいらないかなって思いました。

rira100000000 replied:
こちらもメソッドを削除しました。

maedana commented on 05.wc/lib/wc.rb:
```diff
@@ -0,0 +1,122 @@
+      count_list[index][:number_of_lines], count_list[index][:number_of_words], count_list[index][:number_of_bytes] = *return_zeros_for_counts
+    elsif File.exist?(path)
+      count_list[index][:number_of_lines], count_list[index][:number_of_words], count_list[index][:number_of_bytes] = *return_count_list(path)
+      total_hash[:lines_total], total_hash[:words_total], total_hash[:bytes_total] = *return_total_list(total_hash, count_list, index)
```
ここでやりたいことって結局は`return_count_list`の値を`total_hash`の各値に足していきたいだけですよね。メソッドにせずともここでシンプルに足し込んでいくのでもいいのかなって思いました。

今のままメソッド内で足し込んだ値を返すでもいいと思いますが、少なくとも引数は`total_hash`と`count_list[index]`の2つに減らすことは出来るかなと思いましたー

rira100000000 replied:
メソッドを削除しました。

削除した結果、rubocopの行数制限に引っかかるようになったのでループの処理をまとめてanalyse_pathsメソッドに分離しました。

rira100000000 committed:
メソッド名の変更

return_count_listをlist_countsに変更しました

rira100000000 committed:
不要なメソッドの削除

return_zeros_for_countsを削除しました

rira100000000 committed:
不要なメソッドの削除

return_total_listメソッドを削除しました

rira100000000 committed:
Rubocop対応

行数制限に対応するため、print_command_line_argument_countの処理の一部をanalyse_pathsに分離しました
